File: markwon-inline-parser/src/main/java/io/noties/markwon/inlineparser/CloseBracketInlineProcessor.java
package io.noties.markwon.inlineparser;

import static io.noties.markwon.inlineparser.InlineParserUtils.mergeChildTextNodes;

import org.commonmark.internal.Bracket;
import org.commonmark.internal.util.Escaping;
import org.commonmark.node.Image;
import org.commonmark.node.Link;
import org.commonmark.node.LinkReferenceDefinition;
import org.commonmark.node.Node;
import org.commonmark.parser.beta.Position;

/**
 * Parses markdown link or image, relies on {@link OpenBracketInlineProcessor}
 * to handle start of these elements
 *
 * @since 4.2.0
 */
public class CloseBracketInlineProcessor extends InlineProcessor {

    @Override
    public char specialCharacter() {
        return ']';
    }

    @Override
    protected Node parse() {
        scanner.next();
        Position closeBracketPosition = scanner.position();

        // Get previous `[` or `![`
        Bracket opener = lastBracket();
        if (opener == null) {
            // No matching opener, just return a literal.
            return text("]");
        }

        if (!opener.allowed) {
            // Matching opener but it's not allowed, just return a literal.
            removeLastBracket();
            return text("]");
        }
        boolean isImage = opener.markerNode != null;

        String destination = null;
        String title = null;
        boolean isLinkOrImage = false;

        // [text](dest "title")
        if (peek() == '(') {
            scanner.next();
            spnl();

            destination = parseLinkDestination();
            if (destination != null) {
                spnl();
                if (scanner.whitespace() > 0) {
                    title = parseLinkTitle();
                    spnl();
                }
                if (peek() == ')') {
                    scanner.next();
                    isLinkOrImage = true;
                } else {
                    scanner.setPosition(closeBracketPosition);
                }
            }
        }

        // [text][label] یا [text][] یا [text]
        if (!isLinkOrImage) {
            Position beforeLabel = scanner.position();
            parseLinkLabel();
            Position afterLabel = scanner.position();

            int labelLength = scanner.getSource(beforeLabel, afterLabel).getContent().length();
            String reference = null;

            if (labelLength > 2) {
                reference = scanner.getSource(beforeLabel, afterLabel).getContent();
            } else if (!opener.bracketAfter) {
                reference = scanner.getSource(opener.contentPosition, closeBracketPosition).getContent();
            }

            if (reference != null) {
                String normalized = normalizeReference(reference);
                LinkReferenceDefinition def = context.getLinkReferenceDefinition(normalized);
                if (def != null) {
                    destination = def.getDestination();
                    title = def.getTitle();
                    isLinkOrImage = true;
                }
            }
        }

        if (isLinkOrImage) {
            Node linkOrImage = isImage ? new Image(destination, title) : new Link(destination, title);

            Node child = opener.bracketNode.getNext();
            while (child != null && child != opener.bracketNode) {
                Node next = child.getNext();
                linkOrImage.appendChild(child);
                child = next;
            }

            processDelimiters(opener.previousDelimiter);
            mergeChildTextNodes(linkOrImage);

            opener.bracketNode.unlink();
            if (isImage && opener.markerNode != null) {
                opener.markerNode.unlink();
            }
            removeLastBracket();

            // Links داخل لینک ممنوعه
            if (!isImage) {
                Bracket bracket = lastBracket();
                while (bracket != null) {
                    if (bracket.markerNode == null) {
                        bracket.allowed = false;
                    }
                    bracket = bracket.previous;
                }
            }

            return linkOrImage;
        } else {
            scanner.setPosition(closeBracketPosition);
            removeLastBracket();
            return text("]");
        }
    }

    public static String normalizeReference(String input) {
        // "[label]" -> "label"
        String stripped = input.substring(1, input.length() - 1);
        return Escaping.normalizeLabelContent(stripped);
    }
}

File: markwon-inline-parser/src/main/java/io/noties/markwon/inlineparser/CloseBracketInlineProcessor.kt
package io.noties.markwon.inlineparser

import org.commonmark.internal.util.Escaping
import org.commonmark.node.Image
import org.commonmark.node.Link
import org.commonmark.node.Node

/**
 * Parses markdown link or image, relies on [OpenBracketInlineProcessor]
 * to handle start of these elements
 *
 * @since 4.2.0
 */
class CloseBracketInlineProcessor : InlineProcessor() {
    override fun specialCharacter(): Char {
        return ']'
    }

    override fun parse(): Node? {
        scanner.next()
        val closeBracketPosition = scanner.position()

        // Get previous `[` or `![`
        val opener = lastBracket()
        if (opener == null) {
            // No matching opener, just return a literal.
            return text("]")
        }

        if (!opener.allowed) {
            // Matching opener but it's not allowed, just return a literal.
            removeLastBracket()
            return text("]")
        }
        val isImage = opener.markerNode != null

        var destination: String? = null
        var title: String? = null
        var isLinkOrImage = false

        // [text](dest "title")
        if (peek() == '(') {
            scanner.next()
            spnl()

            destination = parseLinkDestination()
            if (destination != null) {
                spnl()
                if (scanner.whitespace() > 0) {
                    title = parseLinkTitle()
                    spnl()
                }
                if (peek() == ')') {
                    scanner.next()
                    isLinkOrImage = true
                } else {
                    scanner.setPosition(closeBracketPosition)
                }
            }
        }

        // [text][label] یا [text][] یا [text]
        if (!isLinkOrImage) {
            val beforeLabel = scanner.position()
            parseLinkLabel()
            val afterLabel = scanner.position()

            val labelLength = scanner.getSource(beforeLabel, afterLabel).getContent().length
            var reference: String? = null

            if (labelLength > 2) {
                reference = scanner.getSource(beforeLabel, afterLabel).getContent()
            } else if (!opener.bracketAfter) {
                reference = scanner.getSource(opener.contentPosition, closeBracketPosition).getContent()
            }

            if (reference != null) {
                val normalized: String? = normalizeReference(reference)
                val def = context.getLinkReferenceDefinition(normalized)
                if (def != null) {
                    destination = def.destination
                    title = def.title
                    isLinkOrImage = true
                }
            }
        }

        if (isLinkOrImage) {
            val linkOrImage = if (isImage) Image(destination, title) else Link(destination, title)

            var child = opener.bracketNode.next
            while (child != null && child !== opener.bracketNode) {
                val next = child.next
                linkOrImage.appendChild(child)
                child = next
            }

            processDelimiters(opener.previousDelimiter)
            InlineParserUtils.mergeChildTextNodes(linkOrImage)

            opener.bracketNode.unlink()
            if (isImage) {
                opener.markerNode.unlink()
            }
            removeLastBracket()

            // Links داخل لینک ممنوعه
            if (!isImage) {
                var bracket = lastBracket()
                while (bracket != null) {
                    if (bracket.markerNode == null) {
                        bracket.allowed = false
                    }
                    bracket = bracket.previous
                }
            }

            return linkOrImage
        } else {
            scanner.setPosition(closeBracketPosition)
            removeLastBracket()
            return text("]")
        }
    }

    companion object {
        fun normalizeReference(input: String): String? {
            // "[label]" -> "label"
            val stripped = input.substring(1, input.length - 1)
            return Escaping.normalizeLabelContent(stripped)
        }
    }
}
